server/src/common/server.ts

```diff
@@ -1210,6 +1218,18 @@ export interface _Connection<PConsole = _, PTracer = _, PTelemetry = _, PClient
 	 */
 	onDidSaveTextDocument(handler: NotificationHandler<DidSaveTextDocumentParams>): void;
 
+	/**
+	 * Installs a handler for the `WillCreateFiles` request.
+	 *
+	 * Note that this request is opt-in. The client will not send it unless
+	 * your server has the `files.willCreate` capability,
+	 * or you've dynamically registered for the `workspace/willCreateFiles`
+	 * method.
+	 *
+	 * @param handler The corresponding handler.
+	 */
+	onWillCreateFiles(handler: RequestHandler<CreateFilesParams, WorkspaceEdit | undefined | null, void>): void;
```

Comment:
I moved away from putting the handlers on the connection namespace since it makes it hard to understand this. Since we already have a `window` namespace and that namespace is use for UI specific requests and notification we should add these to the `window` namespace. This makes it also clear that this has nothing to do with normal file events since those would be on the workspace namespace. All window stuff is UI related.

Reply:
This also allows you to implement this on the server side as a feature and not directly in the server since the feature can be added to the RemoteWindow type here https://github.com/microsoft/vscode-languageserver-node/blob/master/server/src/common/server.ts#L470 and to the implementation here: https://github.com/microsoft/vscode-languageserver-node/blob/master/server/src/common/server.ts#L512

I can help with this if it is not clear how to do this.

Reply:
I started doing this, but I've gotten confused. The example I was copying (ShowDocumentFeature) is a server->client request. The ones I'm implementing come from the client. So I don't have a server implementation that's shared, the implementation needs to come from the specific server implementation.

I've pushed my changes so far, there are two places I'm stuck:

1. What the (test) server implementation should look like now (this code doesn't compile):

https://github.com/microsoft/vscode-languageserver-node/pull/687/files#diff-18e38934adbfa54d730ee13bc65f20e81c71d6b5fcc526f1a6c543d9cb2995dbR243

2. What should the Feature look like (when the server code has no default implementation unless it's being contributed by the implementing server):

https://github.com/microsoft/vscode-languageserver-node/pull/687/files#diff-36df91f0cf05683eb4b817da9fe0d3f54e6110d749be36546e3e47ac09f72280R15

Reply:
You can copy from an example that implements a language request. It would work the same for the window namespace. A corresponding example is here: https://github.com/microsoft/vscode-languageserver-node/blob/master/server/src/common/onTypeRename.ts#L6

The code for the test server is correct assuming that you have it working for the window namespace. Is your code checked in. Then I can tweak it and push to your repository if you give me write access.

Reply:
Aha, I see! I've gotten further (and pushed my changes), though it seems like I'm missing a `this.attachWorkDoneProgress` from `_RemoteWindowImpl` (it's on `_Languages` already). What's the correct way to add this? (I can copy/paste from `_Languages` but I'm not sure if that's right - I'm also unsure if we should have progress here and if so whether I need some tweaks to the spec?).

I've given your access to my fork btw, so feel free to make changes if that's easier than describing. Thanks!

Reply:
OK. I did the changes and pushed them. Did some additional code cleanup One thing I didn't do to not surprise you is to rename protocol.window.fileOperations.ts to protocol.fileOperations. I usually don't put the namespace into the name. A also created a fileOperations.ts file in the client.

Regarding `this.attachWorkDoneProgress` since the request should be fast we shouldn't allow progress. So you use a normal RequestHandler.

Reply:
Great - thanks! I'll rename that file and fill in the rest of the notifications and tests.